extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -89,174 +69,128 @@
       return tasks.get(task.getId()).getResult();
     }
   }
 
-  JobResponse monitorJob(K8sTaskId k8sTaskId)
+  private TaskStatus runTask(Task task)
   {
-    return monitorJob(client.getMainJobPod(k8sTaskId), k8sTaskId);
+    return doTask(task, true);
   }
 
-  JobResponse monitorJob(Pod peonPod, K8sTaskId k8sTaskId)
+  private TaskStatus joinTask(Task task)
   {
-    if (peonPod == null) {
-      throw new ISE("Error in k8s launching peon pod for task %s", k8sTaskId);
-    }
-    return client.waitForJobCompletion(
-        k8sTaskId,
-        KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.maxTaskDuration),
-        TimeUnit.MILLISECONDS
-    );
+    return doTask(task, false);
   }
 
-  private TaskStatus getTaskStatus(K8sTaskId task, JobResponse jobResponse) throws IOException
+  @VisibleForTesting
+  protected TaskStatus doTask(Task task, boolean run)
   {
-    Optional<InputStream> maybeTaskStatusStream = taskLogs.streamTaskStatus(task.getOriginalTaskId());
-    if (maybeTaskStatusStream.isPresent()) {
-      String taskStatus = IOUtils.toString(maybeTaskStatusStream.get(), StandardCharsets.UTF_8);
-      return mapper.readValue(taskStatus, TaskStatus.class);
-    } else if (PeonPhase.SUCCEEDED.equals(jobResponse.getPhase())) {
-      // fallback to behavior before the introduction of task status streaming for backwards compatibility
-      return TaskStatus.success(task.getOriginalTaskId());
-    } else if (Objects.isNull(jobResponse.getJob())) {
-      return TaskStatus.failure(
-          task.getOriginalTaskId(),
-          StringUtils.format("Task [%s] failed kubernetes job disappeared before completion", task.getOriginalTaskId())
-      );
-    } else {
-      return TaskStatus.failure(
-          task.getOriginalTaskId(),
-          StringUtils.format("Task [%s] failed", task.getOriginalTaskId())
-      );
+    KubernetesPeonLifecycle peonLifecycle = peonLifecycleFactory.build(task);
+
+    synchronized (tasks) {
+      KubernetesWorkItem workItem = tasks.get(task.getId());
+
+      if (workItem == null) {
+        throw new ISE("Task [%s] disappeared", task.getId());
+      }
+
+      if (workItem.isShutdownRequested()) {
+        throw new ISE("Task [%s] has been shut down", task.getId());
+      }
+
+      workItem.setKubernetesPeonLifecycle(peonLifecycle);
+    }
+
+    try {
+      TaskStatus taskStatus;
+      if (run) {
+        taskStatus = peonLifecycle.run(
+            adapter.fromTask(task),
+            config.getTaskLaunchTimeout().toStandardDuration().getMillis(),
+            config.getTaskTimeout().toStandardDuration().getMillis()
+        );
+      } else {
+        taskStatus = peonLifecycle.join(
+            config.getTaskTimeout().toStandardDuration().getMillis()
+        );
+      }
+
+      updateStatus(task, taskStatus);
+
+      return taskStatus;
+    }
+
+    catch (Exception e) {
+      log.error(e, "Task [%s] execution caught an exception", task.getId());
```

Comment:
do we cleanup the task when such an exception happens?

Reply:
In the join method of the KubernetesPeonLifecycle class, there's a finally block that cleans up the K8s Job. this will get run even if the main try logic throws an error.


if there's some problem with this logic in the lifecycle class (like k8s being offline or something), there is a cleanup executor that delete jobs that have been around for a while